execution/inc/leatherman/execution/execution.hpp

```diff
@@ -0,0 +1,311 @@
+     */
+    struct child_exit_exception : execution_failure_exception
+    {
+        /**
+         * Constructs a child_exit_exception.
+         * @param message The exception message.
+         * @param status_code The exit status code of the child process.
+         * @param output The child process stdout output.
+         * @param error The child process stderr output.
+         */
+        child_exit_exception(std::string const& message, int status_code, std::string output, std::string error);
+
+        /**
+         * Gets the child process exit status code.
+         * @return Returns the child process exit status code.
+         */
+        int status_code() const;
+
+     private:
+        int _status_code;
+    };
+
+    /**
+     * Exception that is thrown when a child exists due to a signal.
+     */
+    struct child_signal_exception : execution_failure_exception
+    {
+        /**
+         * Constructs a child_signal_exception.
+         * @param message The exception message.
+         * @param signal The signal code that terminated the child process.
+         * @param output The child process stdout output.
+         * @param error The child process stderr output.
+         */
+        child_signal_exception(std::string const& message, int signal, std::string output, std::string error);
+
+        /**
+         * Gets the signal that terminated the child process.
+         * @return Returns the signal that terminated the child process.
+         */
+        int signal() const;
+
+     private:
+        int _signal;
+    };
+
+    /**
+     * Exception that is thrown when a command times out.
+     */
+    struct timeout_exception : execution_exception
+    {
+        /**
+         * Constructs a timeout_exception.
+         * @param message The exception message.
+         * @param pid The process id of the process that timed out and was killed.
+         */
+        timeout_exception(std::string const& message, size_t pid);
+
+        /**
+         * Gets the process id of the process that timed out and was killed.
+         * @return Returns the process id of the process that timed out and was killed.
+         */
+        size_t pid() const;
+
+     private:
+        size_t _pid;
+    };
+
+    /**
+     * Searches the given paths for the given executable file.
+     * @param file The file to search for.
+     * @param directories The directories to search.
+     * @return Returns the full path or empty if the file could not be found.
+     */
+    std::string which(std::string const& file, std::vector<std::string> const& directories = lth_util::environment::search_paths());
+
+    /**
+     * Expands the executable in the command to the full path.
+     * @param command The command to expand.
+     * @param directories The directories to search.
+     * @return Returns the expanded command if the executable was found or the original command if not.
```

Comment:
This comment changed in https://github.com/puppetlabs/facter/commit/104a9371b98e71aae11965dacff344c26532f8a3.